shotover/src/codec/opensearch.rs

```diff
@@ -0,0 +1,323 @@
+                        self.decode_request(src).unwrap() // TODO
+                    } else {
+                        self.decode_response(src).unwrap() // TODO
+                    };
+
+                    if let Some((amt, parts, cl)) = decode_res {
+                        self.state = State::ReadingBody(parts, cl);
+                        src.advance(amt);
+                    } else {
+                        return Ok(None);
+                    };
+                }
+                State::ReadingBody(parts, cl) => {
+                    if src.is_empty() {
+                        // sometimes messages will have a content-length but no
+                        // body ?
+                        return Ok(Some(vec![Message::from_frame(Frame::OpenSearch(
+                            OpenSearchFrame {
+                                headers: Arc::new(parts),
+                                body: bytes::Bytes::new(),
+                            },
+                        ))]));
+                    }
+
+                    if src.len() < cl {
+                        self.state = State::ReadingBody(parts, cl);
+                        return Ok(None);
+                    }
+
+                    let body = src.split_to(cl).freeze();
+                    return Ok(Some(vec![Message::from_frame(Frame::OpenSearch(
+                        OpenSearchFrame {
+                            headers: Arc::new(parts),
+                            body,
+                        },
+                    ))]));
+                }
+            }
+        }
+    }
+}
+
+pub struct OpenSearchEncoder {
+    direction: Direction,
+}
+
+impl OpenSearchEncoder {
+    pub fn new(direction: Direction) -> Self {
+        Self { direction }
+    }
+}
+
+impl Encoder<Messages> for OpenSearchEncoder {
+    type Error = CodecWriteError;
+
+    fn encode(
+        &mut self,
+        item: Messages,
+        dst: &mut BytesMut,
+    ) -> std::result::Result<(), Self::Error> {
+        item.into_iter().try_for_each(|m| {
+            let start = dst.len();
+            m.ensure_message_type(MessageType::OpenSearch)
+                .map_err(CodecWriteError::Encoder)?;
+            let result = match m.into_encodable() {
+                Encodable::Bytes(bytes) => {
+                    dst.extend_from_slice(&bytes);
+                    Ok(())
+                }
+                Encodable::Frame(frame) => {
+                    let opensearch_frame = frame.into_opensearch().unwrap();
+                    match opensearch_frame.headers.as_ref() {
+                        HttpHead::Request(request_parts) => {
+                            dst.extend_from_slice(
+                                format!("{} ", request_parts.method.as_str()).as_bytes(),
+                            );
+                            dst.extend_from_slice(format!("{} ", request_parts.uri).as_bytes());
+                            dst.extend_from_slice(b"HTTP/1.1");
```

Comment:
what if we received a request that was a different version of HTTP?
Cant we use a crate for HTTP encoding?

Reply:
We only accept HTTP/1 in the decode side of the codec. OpenSearch only started supporting HTTP/2 last year so I think it's safe to only support HTTP/1 for the moment.